pkg/k8s/version/version.go

```diff
@@ -137,21 +162,94 @@ func fallbackDiscovery(client kubernetes.Interface) error {
+		// disabled and the agent can continue.
+		return nil
 	}
 
 	// Unknown error, we can't derive whether to enable or disable
-	// EndpointSlices and need to error out
+	// EndpointSlices and need to error out.
 	return fmt.Errorf("unable to validate EndpointSlices support: %s", err)
 }
 
+func leasesFallbackDiscovery(client kubernetes.Interface, conf k8sconfig.Configuration) error {
+	// K8sEnableLeasesFallbackDiscovery is used to fallback leases discovery to directly
+	// probing the API when we cannot discover API groups.
+	// We require to check for Leases capabilities in operator only, which uses Leases
+	// for leader election purposes in HA mode.
+	if !conf.K8sLeasesFallbackDiscoveryEnabled() {
+		log.Debugf("Skipping Leases support fallback discovery")
+		return nil
+	}
+
+	cached.mutex.RLock()
+	// Here we check if cilium has required authorization to check for leases object or not.
+	// Prior to v1.14 leases.coordination.k8s.io/v1 was not supported and hence cilium
+	// ClusterRole did not had access to corresponding API resource.
+	// Look at the helm manifest to check the ClusterRole associated with cilium Deployments
+	// corresponding to the configured Kubernetes version.
+	if !isGEThanLeaseSupportConstraint(cached.version) {
+		cached.mutex.RUnlock()
+		return nil
+	}
+	cached.mutex.RUnlock()
+
+	// Similar to endpointSlicesFallbackDiscovery here we fallback to probing the Kubernetes
+	// API directly. `kube-controller-manager` creates a lease in the kube-system namespace
+	// and here we try and see if that Lease exists.
+	_, err := client.CoordinationV1().Leases("kube-system").Get(context.TODO(), "kube-controller-manager", metav1.GetOptions{})
+	if err == nil {
+		cached.mutex.Lock()
+		cached.capabilities.LeasesResourceLock = true
+		cached.mutex.Unlock()
+		return nil
+	}
+
+	if errors.IsNotFound(err) {
+		log.WithError(err).Info("Unable to retrieve Leases for kube-controller-manager. Disabling LeasesResourceLock")
+		// StatusNotFound is a safe error, Leases are
+		// disabled and the agent can continue
+		return nil
+	}
+
+	// Unknown error, we can't derive whether to enable or disable
+	// LeasesResourceLock and need to error out
+	return fmt.Errorf("unable to validate LeasesResourceLock support: %s", err)
+}
+
+func updateK8sServerVersion(client kubernetes.Interface) error {
+	sv, err := client.Discovery().ServerVersion()
+	if err != nil {
+		return err
+	}
+
+	// Try GitVersion first. In case of error fallback to MajorMinor
+	if sv.GitVersion != "" {
+		// This is a string like "v1.9.0"
+		ver, err := versioncheck.Version(sv.GitVersion)
+		if err == nil {
+			updateVersion(ver)
+			return nil
+		}
+	}
+
+	if sv.Major != "" && sv.Minor != "" {
+		ver, err := versioncheck.Version(fmt.Sprintf("%s.%s", sv.Major, sv.Minor))
+		if err == nil {
+			updateVersion(ver)
+			return nil
+		}
+	}
+
+	if err != nil {
+		return fmt.Errorf("Cannot parse k8s server version from %+v: %s", sv, err)
```

Comment:
Could you please ensure that all errors are formatted starting with a lower-case?